File: samples/swatch.c
/* Render the baseline dark accents as ANSI swatches with WCAG contrast. */
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#define SURFACE "#111418"
#define MIN_RATIO 4.5 /* WCAG AA for normal text */

typedef struct {
    const char *name;
    const char *hex;
} swatch;

static const swatch accents[] = {
    {"green", "#A1D39A"},   {"blue", "#AAC7FF"}, {"orange", "#FDB975"},
    {"cyan", "#83D3E3"},    {"magenta", "#EBB5ED"},
    {"yellow", "#D4C871"},  {"red", "#FFB3AC"},
};

static int parse_rgb(const char *hex, unsigned *r, unsigned *g, unsigned *b)
{
    return sscanf(hex, "#%02x%02x%02x", r, g, b) == 3;
}

static double linear(unsigned channel)
{
    double v = channel / 255.0;
    return v <= 0.04045 ? v / 12.92 : pow((v + 0.055) / 1.055, 2.4);
}

static double luminance(unsigned r, unsigned g, unsigned b)
{
    return 0.2126 * linear(r) + 0.7152 * linear(g) + 0.0722 * linear(b);
}

static double contrast(double a, double b)
{
    double hi = a > b ? a : b;
    double lo = a > b ? b : a;
    return (hi + 0.05) / (lo + 0.05);
}

static const char *grade(double ratio)
{
    if (ratio >= 7.0)
        return "AAA";
    if (ratio >= MIN_RATIO)
        return "AA";
    if (ratio >= 3.0)
        return "AA-LARGE";
    return "FAIL";
}

int main(void)
{
    unsigned r, g, b;
    if (!parse_rgb(SURFACE, &r, &g, &b)) {
        fprintf(stderr, "bad surface: %s\n", SURFACE);
        return EXIT_FAILURE;
    }
    double background = luminance(r, g, b);
    for (size_t i = 0; i < sizeof accents / sizeof *accents; i++) {
        if (!parse_rgb(accents[i].hex, &r, &g, &b)) {
            fprintf(stderr, "%s: bad hex %s\n", accents[i].name, accents[i].hex);
            return EXIT_FAILURE;
        }
        double ratio = contrast(luminance(r, g, b), background);
        printf("\x1b[48;2;%u;%u;%um    \x1b[0m %-8s %s  %5.2f:1  %s\n",
               r, g, b, accents[i].name, accents[i].hex, ratio, grade(ratio));
    }
    return EXIT_SUCCESS;
}

File: samples/swatch.go
// Command swatch renders the baseline dark accents as ANSI swatches with
// their WCAG contrast against the theme surface.
package main

import (
	"fmt"
	"math"
	"os"
)

const surface = "#111418"

// minRatio is WCAG AA for normal text.
const minRatio = 4.5

type swatch struct {
	name, hex string
}

var accents = []swatch{
	{"green", "#A1D39A"},
	{"blue", "#AAC7FF"},
	{"orange", "#FDB975"},
	{"cyan", "#83D3E3"},
	{"magenta", "#EBB5ED"},
	{"yellow", "#D4C871"},
	{"red", "#FFB3AC"},
}

func (s swatch) rgb() (r, g, b int, err error) {
	if len(s.hex) != 7 || s.hex[0] != '#' {
		return 0, 0, 0, fmt.Errorf("not a hex color: %q", s.hex)
	}
	_, err = fmt.Sscanf(s.hex, "#%02x%02x%02x", &r, &g, &b)
	return r, g, b, err
}

func luminance(r, g, b int) float64 {
	linear := func(channel int) float64 {
		v := float64(channel) / 255
		if v <= 0.04045 {
			return v / 12.92
		}
		return math.Pow((v+0.055)/1.055, 2.4)
	}
	return 0.2126*linear(r) + 0.7152*linear(g) + 0.0722*linear(b)
}

func contrast(a, b float64) float64 {
	if b > a {
		a, b = b, a
	}
	return (a + 0.05) / (b + 0.05)
}

func grade(ratio float64) string {
	switch {
	case ratio >= 7:
		return "AAA"
	case ratio >= minRatio:
		return "AA"
	case ratio >= 3:
		return "AA-LARGE"
	default:
		return "FAIL"
	}
}

func main() {
	sr, sg, sb, err := swatch{"surface", surface}.rgb()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	background := luminance(sr, sg, sb)
	for _, s := range accents {
		r, g, b, err := s.rgb()
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
		ratio := contrast(luminance(r, g, b), background)
		fmt.Printf("\x1b[48;2;%d;%d;%dm    \x1b[0m %-8s %s  %5.2f:1  %s\n",
			r, g, b, s.name, s.hex, ratio, grade(ratio))
	}
}
